sql/memo/join_order_builder.go

```diff
@@ -182,6 +180,103 @@ func (j *joinOrderBuilder) populateSubgraph(n sql.Node) (vertexSet, edgeSet, *Ex
 	return j.allVertices().difference(startV), j.allEdges().Difference(startE), group
 }
 
+// ensureClosure adds the closure of all transitive equivalency groups
+// to the join tree. Each transitive edge will add an inner edge, filter,
+// and join group that inherit join type and tree depth from the original
+// join tree.
+func (j *joinOrderBuilder) ensureClosure(grp *ExprGroup) {
+	fds := grp.RelProps.FuncDeps()
+	for _, set := range fds.Equiv().Sets() {
+		for col1, ok1 := set.Next(1); ok1; col1, ok1 = set.Next(col1 + 1) {
+			for col2, ok2 := set.Next(col1 + 1); ok2; col2, ok2 = set.Next(col2 + 1) {
+				if !j.hasEqEdge(col1, col2) {
+					j.makeTransitiveEdge(col1, col2)
+				}
+			}
+		}
+	}
+}
+
+// hasEqEdge returns true if the inner edges include a direct equality between
+// the two given columns (e.g. x = a).
+func (j joinOrderBuilder) hasEqEdge(leftCol, rightCol sql.ColumnId) bool {
+	for idx, ok := j.innerEdges.Next(0); ok; idx, ok = j.innerEdges.Next(idx + 1) {
+		for _, f := range j.edges[idx].filters {
+			var l *ColRef
+			var r *ColRef
+			switch f := f.(type) {
+			case *Equal:
+				l, _ = f.Left.Scalar.(*ColRef)
+				r, _ = f.Right.Scalar.(*ColRef)
+			case *NullSafeEq:
+				l, _ = f.Left.Scalar.(*ColRef)
+				r, _ = f.Right.Scalar.(*ColRef)
+			}
+			if l == nil || r == nil {
+				continue
+			}
+			if (r.Col == leftCol && l.Col == rightCol) ||
+				(r.Col == rightCol && l.Col == leftCol) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// makeTransitiveEdge constructs a new join tree edge and memo group
+// on an equality filter between two columns.
+func (j *joinOrderBuilder) makeTransitiveEdge(col1, col2 sql.ColumnId) {
```

Comment:
Would be great to see unit tests of this function and ensureClosure

Not to test it per se, but to tighten up and make explicit the contract. Basically the best kind of documentation you can write.